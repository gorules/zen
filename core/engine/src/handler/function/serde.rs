use crate::handler::function::error::ResultExt;
use itertools::Itertools;
use rquickjs::{Ctx, FromJs, IntoAtom, IntoJs, Type, Value as QValue};
use serde_json::{json, Map, Number, Value};

#[derive(Debug)]
pub(crate) struct JsValue(pub(crate) Value);

impl<'js> FromJs<'js> for JsValue {
    fn from_js(ctx: &Ctx<'js>, v: QValue<'js>) -> rquickjs::Result<Self> {
        let computed_value = match v.type_of() {
            Type::Uninitialized | Type::Undefined | Type::Null => Value::Null,
            Type::Bool => Value::Bool(v.as_bool().or_throw_msg(ctx, "failed to convert to bool")?),
            Type::Int => Value::Number(Number::from(
                v.as_int().or_throw_msg(ctx, "failed to convert to int")?,
            )),
            Type::BigInt => Value::Number(Number::from(
                v.as_big_int()
                    .map(|b| b.clone().to_i64().ok())
                    .flatten()
                    .or_throw_msg(ctx, "failed to convert to number")?,
            )),
            Type::Float => Value::Number(
                v.as_float()
                    .map(|n| Number::from_f64(n))
                    .flatten()
                    .or_throw_msg(ctx, "failed to convert to number")?,
            ),
            Type::String => Value::String(
                v.as_string()
                    .map(|s| s.to_string().ok())
                    .flatten()
                    .or_throw_msg(ctx, "failed to convert to string")?,
            ),
            Type::Array => {
                let arr = v
                    .as_array()
                    .or_throw_msg(ctx, "failed to convert to array")?;

                let js_arr: Vec<Value> = arr
                    .iter::<QValue>()
                    .map_ok(|n| JsValue::from_js(ctx, n.clone()).map(|js_val| js_val.0))
                    .flatten()
                    .try_collect()
                    .or_throw(ctx)?;

                Value::Array(js_arr)
            }
            Type::Object => {
                let object = v
                    .as_object()
                    .or_throw_msg(ctx, "failed to convert to object")?;

                let js_object: Map<String, Value> = object
                    .props::<String, QValue>()
                    .map_ok(|(key, value)| {
                        JsValue::from_js(ctx, value.clone()).map(|js_val| (key, js_val.0))
                    })
                    .flatten()
                    .try_collect()
                    .or_throw(ctx)?;

                Value::Object(js_object)
            }
            Type::Exception => {
                let exception = v
                    .as_exception()
                    .or_throw_msg(ctx, "failed to convert to exception")?;

                let message = exception.message().unwrap_or_default();
                let description = exception.to_string();

                json!({ "message": message, "description": description })
            }
            Type::Function => json!("[Function]"),
            Type::Module => json!("[Module]"),
            Type::Constructor => json!("[Constructor]"),
            Type::Symbol => json!("[Symbol]"),
            Type::Unknown => json!("[Unknown]"),
            Type::Promise => {
                let promise = v.as_promise().or_throw(ctx)?;
                let val: JsValue = promise.finish()?;
                val.0
            }
        };

        Ok(JsValue(computed_value))
    }
}

impl<'js> IntoJs<'js> for JsValue {
    fn into_js(self, ctx: &Ctx<'js>) -> rquickjs::Result<QValue<'js>> {
        let res = match self.0 {
            Value::Null => QValue::new_null(ctx.clone()),
            Value::Bool(b) => QValue::new_bool(ctx.clone(), b),
            Value::Number(n) => QValue::new_number(
                ctx.clone(),
                n.as_f64()
                    .or_throw_msg(ctx, "failed to convert float to number")?,
            ),
            Value::String(str) => str.into_js(ctx)?,
            Value::Array(arr) => {
                let qarr = rquickjs::Array::new(ctx.clone())?;
                for (idx, item) in arr.into_iter().enumerate() {
                    qarr.set(idx, JsValue(item))?;
                }

                qarr.into_value()
            }
            Value::Object(map) => {
                let qmap = rquickjs::Object::new(ctx.clone())?;
                for (key, value) in map.into_iter() {
                    qmap.set(key.into_atom(ctx)?, JsValue(value))?;
                }

                qmap.into_value()
            }
        };

        Ok(res)
    }
}
