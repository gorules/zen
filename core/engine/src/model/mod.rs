use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[cfg(feature = "bincode")]
use bincode::{Decode, Encode};

/// JDM Decision model
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct DecisionContent {
    pub nodes: Vec<DecisionNode>,
    pub edges: Vec<DecisionEdge>,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct DecisionEdge {
    pub source_id: String,
    pub target_id: String,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct DecisionNode {
    pub id: String,
    pub name: String,
    #[serde(rename = "type")]
    #[serde(flatten)]
    pub kind: DecisionNodeKind,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(tag = "type")]
#[serde(rename_all = "camelCase")]
pub enum DecisionNodeKind {
    InputNode,
    OutputNode,
    FunctionNode { content: String },
    DecisionNode { content: DecisionNodeContent },
    DecisionTableNode { content: DecisionTableContent },
    ExpressionNode { content: ExpressionNodeContent },
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct DecisionNodeContent {
    pub key: String,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct DecisionTableContent {
    pub rules: Vec<HashMap<String, String>>,
    pub inputs: Vec<DecisionTableInputField>,
    pub outputs: Vec<DecisionTableOutputField>,
    pub hit_policy: DecisionTableHitPolicy,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub enum DecisionTableHitPolicy {
    First,
    Collect,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct DecisionTableInputField {
    pub id: String,
    pub name: String,
    pub field: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct DecisionTableOutputField {
    pub id: String,
    pub name: String,
    pub field: String,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct ExpressionNodeContent {
    pub expressions: Vec<Expression>,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct Expression {
    pub id: String,
    pub key: String,
    pub value: String,
}
