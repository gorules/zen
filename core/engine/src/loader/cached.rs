use std::collections::HashMap;
use std::future::Future;
use std::sync::Arc;

use tokio::sync::Mutex;

use crate::loader::{DecisionLoader, LoaderResponse};
use crate::model::DecisionContent;

pub struct CachedLoader<Loader: DecisionLoader + 'static> {
    loader: Arc<Loader>,
    cache: Mutex<HashMap<String, Arc<DecisionContent>>>,
}

impl<Loader: DecisionLoader + 'static> From<Arc<Loader>> for CachedLoader<Loader> {
    fn from(value: Arc<Loader>) -> Self {
        Self {
            loader: value,
            cache: Mutex::new(HashMap::new()),
        }
    }
}

impl<Loader: DecisionLoader + 'static> DecisionLoader for CachedLoader<Loader> {
    fn load<'a>(&'a self, key: &'a str) -> impl Future<Output = LoaderResponse> + 'a {
        async move {
            let mut cache = self.cache.lock().await;
            if let Some(content) = cache.get(key) {
                return Ok(content.clone());
            }

            let decision_content = self.loader.load(key).await?;
            cache.insert(key.to_string(), decision_content.clone());
            Ok(decision_content)
        }
    }
}
