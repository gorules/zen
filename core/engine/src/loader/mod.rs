mod closure;
mod filesystem;
mod memory;
mod noop;

pub use closure::ClosureLoader;
pub use filesystem::{FilesystemLoader, FilesystemLoaderOptions};
pub use memory::MemoryLoader;
pub use noop::NoopLoader;

use async_trait::async_trait;

use crate::model::DecisionContent;
use std::fmt::Debug;
use std::sync::Arc;
use thiserror::Error;

pub type LoaderResult<T> = Result<T, Box<LoaderError>>;
pub type LoaderResponse = LoaderResult<Arc<DecisionContent>>;

/// Trait used for implementing a loader for decisions
#[async_trait]
pub trait DecisionLoader {
    async fn load(&self, key: &str) -> LoaderResponse;
}

#[derive(Error, Debug)]
pub enum LoaderError {
    #[error("Loader did not find item with key {0}")]
    NotFound(String),
    #[error("Loader failed internally on key {key}: {source}.")]
    Internal {
        key: String,
        #[source]
        source: anyhow::Error,
    },
}
