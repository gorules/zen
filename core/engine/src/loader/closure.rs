use crate::loader::{DecisionLoader, LoaderResponse};
use async_trait::async_trait;
use std::future::Future;

/// Loads decisions using an async closure
#[derive(Debug)]
pub struct ClosureLoader<F>
where
    F: Sync + Send,
{
    closure: F,
}

impl<F, O> ClosureLoader<F>
where
    F: Fn(&str) -> O + Sync + Send,
    O: Future<Output = LoaderResponse> + Send,
{
    pub fn new(closure: F) -> Self {
        Self { closure }
    }
}

#[async_trait]
impl<F, O> DecisionLoader for ClosureLoader<F>
where
    F: Fn(&str) -> O + Sync + Send,
    O: Future<Output = LoaderResponse> + Send,
{
    async fn load(&self, key: &str) -> LoaderResponse {
        let closure = &self.closure;
        closure(key).await
    }
}
