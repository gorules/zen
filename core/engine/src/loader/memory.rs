use crate::loader::{DecisionLoader, LoaderError, LoaderResponse};
use crate::model::DecisionContent;
use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

/// Loads decisions from in-memory hashmap
#[derive(Debug, Default)]
pub struct MemoryLoader {
    memory_refs: RwLock<HashMap<String, Arc<DecisionContent>>>,
}

impl MemoryLoader {
    pub fn add<K, D>(&self, key: K, content: D)
    where
        K: Into<String>,
        D: Into<DecisionContent>,
    {
        let mut mref = self.memory_refs.write().unwrap();
        mref.insert(key.into(), Arc::new(content.into()));
    }

    pub fn get<K>(&self, key: K) -> Option<Arc<DecisionContent>>
    where
        K: AsRef<str>,
    {
        let mref = self.memory_refs.read().unwrap();
        mref.get(key.as_ref()).map(|r| r.clone())
    }

    pub fn remove<K>(&self, key: K) -> bool
    where
        K: AsRef<str>,
    {
        let mut mref = self.memory_refs.write().unwrap();
        mref.remove(key.as_ref()).is_some()
    }
}

#[async_trait]
impl DecisionLoader for MemoryLoader {
    async fn load(&self, key: &str) -> LoaderResponse {
        self.get(&key)
            .ok_or_else(|| LoaderError::NotFound(key.to_string()).into())
    }
}
