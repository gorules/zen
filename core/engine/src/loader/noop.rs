use crate::loader::{DecisionLoader, LoaderError, LoaderResponse};
use anyhow::anyhow;
use async_trait::async_trait;

/// Default loader which always fails
#[derive(Default, Debug)]
pub struct NoopLoader;

#[async_trait]
impl DecisionLoader for NoopLoader {
    async fn load(&self, key: &str) -> LoaderResponse {
        Err(LoaderError::Internal {
            key: key.to_string(),
            source: anyhow!("Loader is no-op"),
        }
        .into())
    }
}
