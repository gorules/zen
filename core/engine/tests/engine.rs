use std::io::Read;
use std::ops::Deref;
use std::path::Path;
use std::sync::Arc;

use serde::Deserialize;
use serde_json::json;

use zen_engine::loader::{LoaderError, MemoryLoader};
use zen_engine::model::DecisionNodeKind;
use zen_engine::{DecisionEngine, EvaluationError, EvaluationOptions};

use crate::support::{create_fs_loader, load_raw_test_data, load_test_data};

mod support;

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_memory_loader() {
    let memory_loader = Arc::new(MemoryLoader::default());
    memory_loader.add("table", load_test_data("table.json"));
    memory_loader.add("function", load_test_data("function.json"));

    let engine = DecisionEngine::new_arc(memory_loader.clone());
    let table = engine.evaluate("table", &json!({ "input": 12 })).await;
    let function = engine.evaluate("function", &json!({ "input": 12 })).await;

    memory_loader.remove("function");
    let not_found = engine.evaluate("function", &json!({})).await;

    assert_eq!(table.unwrap().result, json!({"output": 10}));
    assert_eq!(function.unwrap().result, json!({"output": 24}));
    assert_eq!(not_found.unwrap_err().to_string(), "Loader error");
}

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_filesystem_loader() {
    let engine = DecisionEngine::new(create_fs_loader());
    let table = engine.evaluate("table.json", &json!({ "input": 12 })).await;
    let function = engine
        .evaluate("function.json", &json!({ "input": 12 }))
        .await;
    let not_found = engine.evaluate("invalid_file", &json!({})).await;

    assert_eq!(table.unwrap().result, json!({"output": 10}));
    assert_eq!(function.unwrap().result, json!({"output": 24}));
    assert_eq!(not_found.unwrap_err().to_string(), "Loader error");
}

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_closure_loader() {
    let engine = DecisionEngine::async_loader(|key| {
        // TODO: Improve once async closures become stable in Rust
        let mv_key = key.to_string();

        async move {
            match mv_key.as_str() {
                "function" => Ok(Arc::new(load_test_data("function.json"))),
                "table" => Ok(Arc::new(load_test_data("table.json"))),
                _ => Err(LoaderError::NotFound(mv_key).into()),
            }
        }
    });

    let table = engine.evaluate("table", &json!({ "input": 12 })).await;
    let function = engine.evaluate("function", &json!({ "input": 12 })).await;
    let not_found = engine.evaluate("invalid_file", &json!({})).await;

    assert_eq!(table.unwrap().result, json!({"output": 10}));
    assert_eq!(function.unwrap().result, json!({"output": 24}));
    assert_eq!(not_found.unwrap_err().to_string(), "Loader error");
}

#[tokio::test]
async fn engine_noop_loader() {
    // Default engine is noop
    let engine = DecisionEngine::default();
    let result = engine.evaluate("any.json", &json!({})).await;

    assert_eq!(result.unwrap_err().to_string(), "Loader error");
}

#[tokio::test]
async fn engine_get_decision() {
    let engine = DecisionEngine::new(create_fs_loader());

    assert!(engine.get_decision("table.json").await.is_ok());
    assert!(engine.get_decision("any.json").await.is_err());
}

#[tokio::test]
async fn engine_create_decision() {
    let engine = DecisionEngine::default();
    engine.create_decision(load_test_data("table.json").into());
}

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_errors() {
    let engine = DecisionEngine::new(create_fs_loader());

    let infinite_fn = engine.evaluate("infinite-function.json", &json!({})).await;
    match infinite_fn.unwrap_err().deref() {
        EvaluationError::NodeError(e) => {
            assert_eq!(e.node_id, "e0fd96d0-44dc-4f0e-b825-06e56b442d78");
            assert_eq!(e.source.to_string(), "Timeout exceeded");
        }
        _ => assert!(false, "Wrong error type"),
    }

    let recursive = engine.evaluate("recursive-table1.json", &json!({})).await;
    match recursive.unwrap_err().deref() {
        EvaluationError::NodeError(e) => {
            assert_eq!(e.source.to_string(), "Depth limit exceeded")
        }
        _ => assert!(false, "Depth limit not exceeded"),
    }
}

#[tokio::test]
async fn engine_with_trace() {
    let engine = DecisionEngine::new(create_fs_loader());

    let table_r = engine.evaluate("table.json", &json!({ "input": 12 })).await;
    let table_opt_r = engine
        .evaluate_with_opts(
            "table.json",
            &json!({ "input": 12 }),
            EvaluationOptions {
                trace: Some(true),
                max_depth: None,
            },
        )
        .await;

    let table = table_r.unwrap();
    let table_opt = table_opt_r.unwrap();

    assert!(table.trace.is_none());
    assert!(table_opt.trace.is_some());

    let trace = table_opt.trace.unwrap();
    assert_eq!(trace.len(), 3); // trace for each node
}

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_function_imports() {
    let mut function_content = load_test_data("function.json");

    let imports_js_path = Path::new("js").join("imports.js");
    let mut replace_buffer = load_raw_test_data(imports_js_path.to_str().unwrap());
    let mut replace_data = String::new();
    replace_buffer.read_to_string(&mut replace_data).unwrap();

    function_content.nodes.iter_mut().for_each(|node| {
        if let DecisionNodeKind::FunctionNode { content } = &mut node.kind {
            let _ = std::mem::replace(content, replace_data.clone());
        }
    });

    let decision = DecisionEngine::default().create_decision(function_content.into());
    let response = decision.evaluate(&json!({})).await.unwrap();

    #[derive(Deserialize, Debug)]
    #[serde(rename_all = "camelCase")]
    struct GraphResult {
        bigjs_tests: Vec<bool>,
        bigjs_valid: bool,
        dayjs_valid: bool,
        moment_valid: bool,
    }

    let result = serde_json::from_value::<GraphResult>(response.result).unwrap();

    assert!(result.bigjs_tests.iter().all(|v| *v));
    assert!(result.bigjs_valid);
    assert!(result.dayjs_valid);
    assert!(result.moment_valid);
}
