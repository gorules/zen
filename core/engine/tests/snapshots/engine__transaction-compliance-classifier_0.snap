---
source: core/engine/tests/engine.rs
expression: serialized_result
---
performance: "[perf]"
result:
  has_compliance_flags: false
  report_instructions: File FinCEN Form 112 within 24 hours
  reporting_deadline:
    "$serde_json::private::Number": "1743330600"
  reporting_status: urgent_filing_required
  requires_immediate_filing: true
trace:
  compliance_flags:
    id: compliance_flags
    input:
      amount_score:
        "$serde_json::private::Number": "15"
      base_risk_score:
        "$serde_json::private::Number": "0"
      classification:
        priority: high_priority
        report_type: CTR
      compliance:
        requires_review: true
        review_reason: Medium risk score with high priority requires review
      customer:
        account_opening_date: 2020-01-15
        high_risk: false
        id: CUST98765
        jurisdiction: US
        kyc_verified: true
        name: Acme Corporation
        normal_activity_patterns:
          - cash_deposit
          - domestic_wire
          - check_deposit
        risk_category: medium
        suspicious_activity_count:
          "$serde_json::private::Number": "0"
        type: business
      customer_score:
        "$serde_json::private::Number": "0"
      priority_score:
        "$serde_json::private::Number": "25"
      reporting:
        days_to_report:
          "$serde_json::private::Number": "15"
        form_name: FinCEN Form 112
      risk:
        jurisdiction_score:
          "$serde_json::private::Number": "15"
      total_risk_score:
        "$serde_json::private::Number": "55"
      transaction:
        amount:
          "$serde_json::private::Number": "15000"
        currency: USD
        date: "2025-03-15T10:30:00Z"
        description: Business cash deposit
        high_risk_country: false
        id: TX123456789
        structuring_suspected: false
        type: cash_deposit
      transaction_history_score:
        "$serde_json::private::Number": "0"
    name: identify_compliance_issues
    order:
      "$serde_json::private::Number": "5"
    output:
      amount_score:
        "$serde_json::private::Number": "15"
      base_risk_score:
        "$serde_json::private::Number": "0"
      classification:
        priority: high_priority
        report_type: CTR
      compliance:
        flags: []
        requires_review: true
        review_reason: Medium risk score with high priority requires review
      customer:
        account_opening_date: 2020-01-15
        high_risk: false
        id: CUST98765
        jurisdiction: US
        kyc_verified: true
        name: Acme Corporation
        normal_activity_patterns:
          - cash_deposit
          - domestic_wire
          - check_deposit
        risk_category: medium
        suspicious_activity_count:
          "$serde_json::private::Number": "0"
        type: business
      customer_score:
        "$serde_json::private::Number": "0"
      priority_score:
        "$serde_json::private::Number": "25"
      reporting:
        days_to_report:
          "$serde_json::private::Number": "15"
        form_name: FinCEN Form 112
      risk:
        jurisdiction_score:
          "$serde_json::private::Number": "15"
      total_risk_score:
        "$serde_json::private::Number": "55"
      transaction:
        amount:
          "$serde_json::private::Number": "15000"
        currency: USD
        date: "2025-03-15T10:30:00Z"
        description: Business cash deposit
        high_risk_country: false
        id: TX123456789
        structuring_suspected: false
        type: cash_deposit
      transaction_history_score:
        "$serde_json::private::Number": "0"
    performance: "[perf]"
    traceData: []
  input:
    id: input
    input: ~
    name: request
    order:
      "$serde_json::private::Number": "0"
    output:
      customer:
        account_opening_date: 2020-01-15
        high_risk: false
        id: CUST98765
        jurisdiction: US
        kyc_verified: true
        name: Acme Corporation
        normal_activity_patterns:
          - cash_deposit
          - domestic_wire
          - check_deposit
        risk_category: medium
        suspicious_activity_count:
          "$serde_json::private::Number": "0"
        type: business
      transaction:
        amount:
          "$serde_json::private::Number": "15000"
        currency: USD
        date: "2025-03-15T10:30:00Z"
        description: Business cash deposit
        high_risk_country: false
        id: TX123456789
        structuring_suspected: false
        type: cash_deposit
    performance: "[perf]"
    traceData: ~
  jurisdiction_rules:
    id: jurisdiction_rules
    input:
      classification:
        priority: high_priority
        report_type: CTR
      customer:
        account_opening_date: 2020-01-15
        high_risk: false
        id: CUST98765
        jurisdiction: US
        kyc_verified: true
        name: Acme Corporation
        normal_activity_patterns:
          - cash_deposit
          - domestic_wire
          - check_deposit
        risk_category: medium
        suspicious_activity_count:
          "$serde_json::private::Number": "0"
        type: business
      transaction:
        amount:
          "$serde_json::private::Number": "15000"
        currency: USD
        date: "2025-03-15T10:30:00Z"
        description: Business cash deposit
        high_risk_country: false
        id: TX123456789
        structuring_suspected: false
        type: cash_deposit
    name: apply_jurisdiction_rules
    order:
      "$serde_json::private::Number": "2"
    output:
      classification:
        priority: high_priority
        report_type: CTR
      customer:
        account_opening_date: 2020-01-15
        high_risk: false
        id: CUST98765
        jurisdiction: US
        kyc_verified: true
        name: Acme Corporation
        normal_activity_patterns:
          - cash_deposit
          - domestic_wire
          - check_deposit
        risk_category: medium
        suspicious_activity_count:
          "$serde_json::private::Number": "0"
        type: business
      reporting:
        days_to_report:
          "$serde_json::private::Number": "15"
        form_name: FinCEN Form 112
      risk:
        jurisdiction_score:
          "$serde_json::private::Number": "15"
      transaction:
        amount:
          "$serde_json::private::Number": "15000"
        currency: USD
        date: "2025-03-15T10:30:00Z"
        description: Business cash deposit
        high_risk_country: false
        id: TX123456789
        structuring_suspected: false
        type: cash_deposit
    performance: "[perf]"
    traceData:
      index:
        "$serde_json::private::Number": "0"
      reference_map:
        classification.report_type: CTR
        customer.jurisdiction: US
      rule:
        _id: r2-1
        "classification.report_type[i2-2]": "'CTR'"
        "customer.jurisdiction[i2-1]": "'US'"
  reporting_requirements:
    id: reporting_requirements
    input:
      amount_score:
        "$serde_json::private::Number": "15"
      base_risk_score:
        "$serde_json::private::Number": "0"
      classification:
        priority: high_priority
        report_type: CTR
      compliance:
        flags: []
        requires_review: true
        review_reason: Medium risk score with high priority requires review
      customer:
        account_opening_date: 2020-01-15
        high_risk: false
        id: CUST98765
        jurisdiction: US
        kyc_verified: true
        name: Acme Corporation
        normal_activity_patterns:
          - cash_deposit
          - domestic_wire
          - check_deposit
        risk_category: medium
        suspicious_activity_count:
          "$serde_json::private::Number": "0"
        type: business
      customer_score:
        "$serde_json::private::Number": "0"
      priority_score:
        "$serde_json::private::Number": "25"
      reporting:
        days_to_report:
          "$serde_json::private::Number": "15"
        form_name: FinCEN Form 112
      risk:
        jurisdiction_score:
          "$serde_json::private::Number": "15"
      total_risk_score:
        "$serde_json::private::Number": "55"
      transaction:
        amount:
          "$serde_json::private::Number": "15000"
        currency: USD
        date: "2025-03-15T10:30:00Z"
        description: Business cash deposit
        high_risk_country: false
        id: TX123456789
        structuring_suspected: false
        type: cash_deposit
      transaction_history_score:
        "$serde_json::private::Number": "0"
    name: determine_reporting_requirements
    order:
      "$serde_json::private::Number": "6"
    output:
      has_compliance_flags: false
      report_instructions: File FinCEN Form 112 within 24 hours
      reporting_deadline:
        "$serde_json::private::Number": "1743330600"
      reporting_status: urgent_filing_required
      requires_immediate_filing: true
    performance: "[perf]"
    traceData:
      has_compliance_flags:
        result: "false"
      report_instructions:
        result: "\"File FinCEN Form 112 within 24 hours\""
      reporting_deadline:
        result: "1743330600"
      reporting_status:
        result: "\"urgent_filing_required\""
      requires_immediate_filing:
        result: "true"
  risk_assessment:
    id: risk_assessment
    input:
      classification:
        priority: high_priority
        report_type: CTR
      customer:
        account_opening_date: 2020-01-15
        high_risk: false
        id: CUST98765
        jurisdiction: US
        kyc_verified: true
        name: Acme Corporation
        normal_activity_patterns:
          - cash_deposit
          - domestic_wire
          - check_deposit
        risk_category: medium
        suspicious_activity_count:
          "$serde_json::private::Number": "0"
        type: business
      reporting:
        days_to_report:
          "$serde_json::private::Number": "15"
        form_name: FinCEN Form 112
      risk:
        jurisdiction_score:
          "$serde_json::private::Number": "15"
      transaction:
        amount:
          "$serde_json::private::Number": "15000"
        currency: USD
        date: "2025-03-15T10:30:00Z"
        description: Business cash deposit
        high_risk_country: false
        id: TX123456789
        structuring_suspected: false
        type: cash_deposit
    name: calculate_risk_score
    order:
      "$serde_json::private::Number": "3"
    output:
      amount_score:
        "$serde_json::private::Number": "15"
      base_risk_score:
        "$serde_json::private::Number": "0"
      classification:
        priority: high_priority
        report_type: CTR
      customer:
        account_opening_date: 2020-01-15
        high_risk: false
        id: CUST98765
        jurisdiction: US
        kyc_verified: true
        name: Acme Corporation
        normal_activity_patterns:
          - cash_deposit
          - domestic_wire
          - check_deposit
        risk_category: medium
        suspicious_activity_count:
          "$serde_json::private::Number": "0"
        type: business
      customer_score:
        "$serde_json::private::Number": "0"
      priority_score:
        "$serde_json::private::Number": "25"
      reporting:
        days_to_report:
          "$serde_json::private::Number": "15"
        form_name: FinCEN Form 112
      risk:
        jurisdiction_score:
          "$serde_json::private::Number": "15"
      total_risk_score:
        "$serde_json::private::Number": "55"
      transaction:
        amount:
          "$serde_json::private::Number": "15000"
        currency: USD
        date: "2025-03-15T10:30:00Z"
        description: Business cash deposit
        high_risk_country: false
        id: TX123456789
        structuring_suspected: false
        type: cash_deposit
      transaction_history_score:
        "$serde_json::private::Number": "0"
    performance: "[perf]"
    traceData:
      amount_score:
        result: "15"
      base_risk_score:
        result: "0"
      customer_score:
        result: "0"
      priority_score:
        result: "25"
      total_risk_score:
        result: "55"
      transaction_history_score:
        result: "0"
  threshold_checks:
    id: threshold_checks
    input:
      amount_score:
        "$serde_json::private::Number": "15"
      base_risk_score:
        "$serde_json::private::Number": "0"
      classification:
        priority: high_priority
        report_type: CTR
      customer:
        account_opening_date: 2020-01-15
        high_risk: false
        id: CUST98765
        jurisdiction: US
        kyc_verified: true
        name: Acme Corporation
        normal_activity_patterns:
          - cash_deposit
          - domestic_wire
          - check_deposit
        risk_category: medium
        suspicious_activity_count:
          "$serde_json::private::Number": "0"
        type: business
      customer_score:
        "$serde_json::private::Number": "0"
      priority_score:
        "$serde_json::private::Number": "25"
      reporting:
        days_to_report:
          "$serde_json::private::Number": "15"
        form_name: FinCEN Form 112
      risk:
        jurisdiction_score:
          "$serde_json::private::Number": "15"
      total_risk_score:
        "$serde_json::private::Number": "55"
      transaction:
        amount:
          "$serde_json::private::Number": "15000"
        currency: USD
        date: "2025-03-15T10:30:00Z"
        description: Business cash deposit
        high_risk_country: false
        id: TX123456789
        structuring_suspected: false
        type: cash_deposit
      transaction_history_score:
        "$serde_json::private::Number": "0"
    name: regulatory_thresholds
    order:
      "$serde_json::private::Number": "4"
    output:
      amount_score:
        "$serde_json::private::Number": "15"
      base_risk_score:
        "$serde_json::private::Number": "0"
      classification:
        priority: high_priority
        report_type: CTR
      compliance:
        requires_review: true
        review_reason: Medium risk score with high priority requires review
      customer:
        account_opening_date: 2020-01-15
        high_risk: false
        id: CUST98765
        jurisdiction: US
        kyc_verified: true
        name: Acme Corporation
        normal_activity_patterns:
          - cash_deposit
          - domestic_wire
          - check_deposit
        risk_category: medium
        suspicious_activity_count:
          "$serde_json::private::Number": "0"
        type: business
      customer_score:
        "$serde_json::private::Number": "0"
      priority_score:
        "$serde_json::private::Number": "25"
      reporting:
        days_to_report:
          "$serde_json::private::Number": "15"
        form_name: FinCEN Form 112
      risk:
        jurisdiction_score:
          "$serde_json::private::Number": "15"
      total_risk_score:
        "$serde_json::private::Number": "55"
      transaction:
        amount:
          "$serde_json::private::Number": "15000"
        currency: USD
        date: "2025-03-15T10:30:00Z"
        description: Business cash deposit
        high_risk_country: false
        id: TX123456789
        structuring_suspected: false
        type: cash_deposit
      transaction_history_score:
        "$serde_json::private::Number": "0"
    performance: "[perf]"
    traceData:
      index:
        "$serde_json::private::Number": "2"
      reference_map:
        classification.priority: high_priority
        total_risk_score:
          "$serde_json::private::Number": "55"
      rule:
        _id: r3-3
        "classification.priority[i3-2]": "'high_priority'"
        "total_risk_score[i3-1]": "[40..60]"
  transaction_type:
    id: transaction_type
    input:
      customer:
        account_opening_date: 2020-01-15
        high_risk: false
        id: CUST98765
        jurisdiction: US
        kyc_verified: true
        name: Acme Corporation
        normal_activity_patterns:
          - cash_deposit
          - domestic_wire
          - check_deposit
        risk_category: medium
        suspicious_activity_count:
          "$serde_json::private::Number": "0"
        type: business
      transaction:
        amount:
          "$serde_json::private::Number": "15000"
        currency: USD
        date: "2025-03-15T10:30:00Z"
        description: Business cash deposit
        high_risk_country: false
        id: TX123456789
        structuring_suspected: false
        type: cash_deposit
    name: classify_transaction
    order:
      "$serde_json::private::Number": "1"
    output:
      classification:
        priority: high_priority
        report_type: CTR
      customer:
        account_opening_date: 2020-01-15
        high_risk: false
        id: CUST98765
        jurisdiction: US
        kyc_verified: true
        name: Acme Corporation
        normal_activity_patterns:
          - cash_deposit
          - domestic_wire
          - check_deposit
        risk_category: medium
        suspicious_activity_count:
          "$serde_json::private::Number": "0"
        type: business
      transaction:
        amount:
          "$serde_json::private::Number": "15000"
        currency: USD
        date: "2025-03-15T10:30:00Z"
        description: Business cash deposit
        high_risk_country: false
        id: TX123456789
        structuring_suspected: false
        type: cash_deposit
    performance: "[perf]"
    traceData:
      index:
        "$serde_json::private::Number": "0"
      reference_map:
        transaction.amount:
          "$serde_json::private::Number": "15000"
        transaction.type: cash_deposit
      rule:
        _id: r1-1
        "transaction.amount[i1-2]": "> 10000"
        "transaction.type[i1-1]": "'cash_deposit'"
